cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,473 @@
+  * Use the Cilium Identity to look up TLS details.
+  * Special case: peer node may be the same node as the current node. In this case, the Node can connect to its own serving port, and perform the handshake, just like a remote node.
+  * Validate certificate of the peer. Note that since we are also planning to handle re-key events by pre-emptively re-authenticating, this should not occur very often.
+  Intial implementation efforts show that when using SPIRE, the SPIRE server will pre-emptively re-key for you before the expiration time has elapsed, and send a new keypair via the DelegatedIdentity API - we are planning on triggering reauth when this occurs.
+    * Q: How to make invalid certificate errors super obvious for debugging?
+      * Logs
+      * Metrics
+
+##### Authenticated session established with peer
+
+* Update datapath to inform it that this peer is authenticated (see Data Plane for details)
+* The destination node is responsible for checking that a Pod matching the destination Identity is running on itself, and only allowing the authentication to succeed if it is.
+
+##### Endpoint deletion
+
+* Node with endpoint must clear out Auth table to remove any state that implies authentication occurred for connections with this endpoint
+* Do all other nodes need to clear out their corresponding state also?
+  * Thinking of IP recycling case, e.g. in ENI mode an IP could move from one node to another.
+  * Trigger on CiliumEndpoint deletion even for CiliumEndpoints on remote nodes, delete authentication state corresponding to that peer. Including userspace authentication table entries, datapath CT/map entries.
+
+##### Timer-based
+* Do we need periodic garbage collection, e.g. of authenticated sessions? Because the auth table has an expiry time, we have built-in garbage collection - the userspace will be responsible for pruning expired auth table entries if the identity has been deleted and the certificate validity period has passed.
+* At this stage, it seems likely that deletion of expired entries will still be done by userspace, although this could change in the future.
+
+#### Certificate management
+
+Initial certificate management for this feature will be using SPIRE and SPIFFE, with interfaces built to allow us to walk back to other options once we have more user intent data.
+
+### Data Plane
+
+
+#### Structures
+Policy Entry - 1 bit: "Authentication Required"
+Drop events - 1 codepoint - "Authentication Required"
+Dedicated bpf auth map, see Key question: How to implement dataplane auth check
+
+#### Logic
+When performing policy lookup, check 'authentication required' bit.
+If this bit is set and there is no corresponding entry in the BPF auth table, drop the traffic and emit an 'authentication required' event to the local control plane.
+Ensure both egress and ingress implement this logic.
+For source and destination on the same node, the node will connect to itself and perform a handshake.
```

Comment:
nit: I think this is accurate, but this is out-of-scope for the datapath section. The datapath can't perform this piece, that's already prescribed in the "Triggered by datapath drop" event under control plane above.

```suggestion
```